### DISCUSSION TITLE
Configuring multithreaded software decoding supported?

### DISCUSSION THREAD

**Discussion by basisbit** [OvenMediaLabs/OvenMediaEngine#1548] [Help]
Does OME support configuring it to use multiple CPU threads when decoding an incoming video stream which it can't handle with a single CPU thread? For example I am currently struggling with making OME support transcoding an incoming h265 1440p video stream on a 8 or 16 or 32 cores CPU, and it is mostly idle except for a single thread being at 100% for the decoding.

As an example, assume I want either the full original stream to be delivered over HLS or alternatively a 360p 1Mb/s x264 video stream as fallback.

Currently I see something like this in htop with the named threads view of ome enabled:
![image](https://github.com/AirenSoft/OvenMediaEngine/assets/10870269/7fcd4f05-d258-4a61-9f17-2ce1466e778f)

**Comment by bchah** [comment]
+1 million for this! There are certain features I advise users to disable in my app based on whether they're likely to overwhelm this single thread with an image encoder.

**Comment by Keukhan** [comment]
@basisbit Thanks for reporting

Basically, when decoding one video stream, the number of threads is automatically created according to the number of CPU cores. I need to analyze the reason why load balancing between threads is not possible.

I have one question. Is this issue only happening with HEVC? Does H264 have evenly distributed thread usage?

**Comment by Keukhan** [comment]
@basisbit 

The name and number of decoder threads are not accurate. This is because it follows the name of the thread that opens the codec. At top, the name InboundWorker is the actual decoder thread.

Anyway, when I tested the HEVC decoding, it was created and decoded in a single thread.
If you modify the /projects/transcoder/codec/decoder/decoder_hevc.cpp file as follows, multi-threaded decoding becomes possible.

![image](https://github.com/AirenSoft/OvenMediaEngine/assets/21077363/fcd53258-5b65-4785-9ba7-1bf8d4d4e24b)


I would like to automatically specify the number of threads if possible, but as the number of threads increases, a delay equal to the number of threads occurs.  I will have to think about whether there is a good way and patch it.
